First fully working version of Brave Talk extension

tackley opened the pull request:
Overhauled from upstream to:
- execlude unsupported platforms (no gcal v1 nor outlook live)
- use a webpack build (and use jquery sourced from npm, not a local file)
- use typescript

And include the ability to create the room on the moderator's behalf.

mrose17 reviewed (approved):
Understood, LGTM.

diracdeltas commented on package.json:
```diff
@@ -0,0 +1,39 @@
+  "description": "Jidesha =======",
+  "main": "index.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1",
```
could you please add some tests for this repo?

diracdeltas replied:
also, given all the dependencies, github actions should be set up to run `npm audit` on every PR.

tackley replied:
github action with build and audit added.

diracdeltas replied:
unresolving because npm test command still needs to be added

tackley replied:
test now added and included in CI

diracdeltas commented:
`npm outdated` shows some outdated dependencies

<img width="703" alt="Screen Shot 2021-12-14 at 3 32 50 PM" src="https://user-images.githubusercontent.com/549654/146096059-bdc04a30-d9df-4f17-8ac1-50dcb23f1671.png">

diracdeltas commented on src/brave-talk.ts:
```diff
@@ -10,3 +10,11 @@ function generateRoomWithoutSeparator() {
 }
+
+export function createRoom(roomUrl: string) {
+  window.open(
```
should probably set `noopener` and `noreferrer` here https://developer.mozilla.org/en-US/docs/Web/API/Window/open

tackley replied:
good suggestion, updated

diracdeltas commented on src/meet-calendar.ts:
```diff
@@ -0,0 +1,688 @@
+      let descriptionContainsURL =
+        this.value &&
+        this.value.length >= 1 &&
+        this.value.indexOf(BASE_URL) !== -1;
```
should use `startsWith` otherwise `https://bad.com/#https://$BASE_URL` would pass this check

tackley replied:
In this specific case, that is the behaviour we want. The description is the large free text box within the calendar entry. This bit of code is checking whether the description has previously been edited to include any kind of `talk.brave.com` url anywhere within it - if so, the description is left alone. This code is not attempting to perform url validation.

If the user (or another extension) has entered `https://bad.com/#https://talk.brave.com` into the description box then it is correct that we should leave it alone. We do not extract the url nor do we open it or follow it.

diracdeltas replied:
got it, thanks

diracdeltas commented on src/meet-calendar.ts:
```diff
@@ -0,0 +1,688 @@
+  /**
+   * Updates meetingId, if there is meetingId set it, if not generate it.
+   */
+  updateMeetingId() {
```
this logic is somewhat complicated and needs some tests

tackley replied:
logic significantly simplified and tests against stub html added

diracdeltas commented on src/meet-calendar.ts:
```diff
@@ -0,0 +1,688 @@
+      }
+
+      return {
+        val: () => element.innerHTML,
```
can you use innerText here instead of innerHTML?

tackley replied:
this code is obsolete and innerText used wherever possible. still one case of HTML addition noted at https://github.com/brave/jidesha/pull/2#pullrequestreview-852123136, which I'll address there.

diracdeltas commented on src/meet-calendar.ts:
```diff
@@ -0,0 +1,688 @@
+  /**
+   * The text from the location input field.
+   */
+  get text(): string | undefined {
```
either this should be renamed to `html()` or https://github.com/brave/jidesha/pull/2/files#diff-0d87d0858bdb97f9e337a8816f11b4c52a42a106faf5c119c89834ba8c783358R414 should be using innerText

tackley replied:
code simplified to address this confusion

diracdeltas commented on src/meet-calendar.ts:
```diff
@@ -0,0 +1,688 @@
+          // and then click join now before saving
+          this.event.updateMeetingId();
+
+          window.open(BASE_URL + this.event.meetingId, "_blank");
```
see https://github.com/brave/jidesha/pull/2/files#r769129574

tackley replied:
updated

diracdeltas commented on src/meet-calendar.ts:
```diff
@@ -0,0 +1,688 @@
+   */
+  updateInitialButtonURL(location: LocationWrapper) {
+    var button = $("#jitsi_button a");
+    button.html("Add a " + LOCATION_TEXT);
```
can you use `.text` or some other safe method (that doesn't potentially inject HTML) for this and other `.html` calls?

tackley replied:
done, in this case and the others where `.html` was being used. Each was just setting plain text.

diracdeltas commented on src/meet-calendar.ts:
```diff
@@ -0,0 +1,688 @@
+ * @param parameterName the name of the param to search for
+ * @returns {String} the parameter value.
+ */
+function findGetParameter(parameterName: string): string | null {
```
can you simplify this with https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams?

tackley replied:
very much so! done.

diracdeltas commented on src/meet-calendar.ts:
```diff
@@ -0,0 +1,688 @@
+      ) {
+        c!.update();
+      }
+    }).observe(body, {
```
we have had some issues in the past with MutationObservers causing page performance issues. you may want to measure page load time before and after this change to make sure it's not a big difference.

tackley replied:
Good suggestion, I will do this.

diracdeltas commented on src/meet-calendar.ts:
```diff
@@ -0,0 +1,688 @@
+ * Checks whether it is ok to add the button to current page
+ * in case of new google calendar interface
+ */
+function checkAndUpdateCalendarG2() {
```
seems like this needs a test, preferably one that loads gcal in puppeteer or some other browser for test purposes.

tackley replied:
tests added in later updates, admittedly using stub html rather as that enables them to be run reliably during the CI build

tackley commented:
> `npm outdated` shows some outdated dependencies
> 
> <img alt="Screen Shot 2021-12-14 at 3 32 50 PM" width="703" src="https://user-images.githubusercontent.com/549654/146096059-bdc04a30-d9df-4f17-8ac1-50dcb23f1671.png">

arggh! that's in the few days since I added them. updated.

tackley commented:
Thanks for the feedback @diracdeltas!

Still outstanding from the points you raise:
- **tests** - wholeheartedly agree that this would be valuable. This would require extensive investment this to run reliably: it does not just require a simple puppeteer setup, as it also requires building and installing the extension as well as interaction with google accounts and services. Furthermore, the extensive use of long term a-b tested rollouts on google products would limit the value of the results. I'll discuss with @mrose17 @mattmcalister @jsecretan on this - I suspect manual testing will be the preference until it is clear how widespread usage of the extension is.
- **complexity/clarity of code** - it's a lot better than it was, but there are indeed still areas of code where the intent is not fully clear, some of which you point out. I was deliberately trying to hold myself back from doing a full rewrite, and leaving code alone that didn't present a risk and worked when tested. Is there anything you are concerned about from a security perspective?
- **jquery** - my knowledge of the security vulnerabilities of jquery are limited. As you see from the code, there is extensive use of appending html-as-text to the dom. Is this acceptable?

diracdeltas commented:
> * **tests** - wholeheartedly agree that this would be valuable. This would require extensive investment this to run reliably: it does not just require a simple puppeteer setup, as it also requires building and installing the extension as well as interaction with google accounts and services. Furthermore, the extensive use of long term a-b tested rollouts on google products would limit the value of the results. I'll discuss with @mrose17 @mattmcalister @jsecretan on this - I suspect manual testing will be the preference until it is clear how widespread usage of the extension is.

I think we need automated tests here, if not with puppeteer then at least some unit tests with stubs/mocks for now. 


The use of jquery seems fine :). Thanks for updating it to append text instead of HTML when possible.

tackley commented:
> my main concern is any potentially untrusted data that gets rendered as html. this should be avoided when possible using safe APIs like `.text` but if absolutely necessary, the input should be sanitized.

so, as a guideline for working with the historical jquery code in this extension:
1. use `.text` rather than `.html` / `.append` whenever possible
2. calling `.html` / `.append` with a constant string (e.g. https://github.com/brave/jidesha/blob/e4ee86f1c1e9ec82feb11f5ba8f24d4a51b969de/src/meet-calendar.ts#L346-L351) is acceptable
3. never call `.html` / `.append` with a constructed string if that string contains un-sanitized user inputs

Looking at the code, the only case where user inputs are potentially mixed into an html call is here: https://github.com/brave/jidesha/blob/e4ee86f1c1e9ec82feb11f5ba8f24d4a51b969de/src/meet-calendar.ts#L490-L510

I'm pretty sure I can avoid that, so the rule I'll follow is:

3. never call `.html` / `.append` with a non-constant string

diracdeltas reviewed (changes requested):
I noticed a bug while testing this out:

1. go to google calendar and click the button to insert a brave talk URL
2. it inserts the URL in the location field but with an extra space at the end 
<img width="556" alt="Screen Shot 2021-12-28 at 2 23 28 PM" src="https://user-images.githubusercontent.com/549654/147611525-28864efd-eeea-483f-9522-4e5002ffe57b.png">

3. save the event, then try clicking the Brave Talk URL in it. instead of going to the URL when clicked, Google Calendar then opens a Google Maps link with the URL as the map location, which leaks the URL to Google Maps

mrose17 commented:
> I noticed a bug while testing this out:
> 
> 1. go to google calendar and click the button to insert a brave talk URL
> 2. it inserts the URL in the location field but with an extra space at the end
> 
> <img alt="Screen Shot 2021-12-28 at 2 23 28 PM" width="556" src="https://user-images.githubusercontent.com/549654/147611525-28864efd-eeea-483f-9522-4e5002ffe57b.png">
> 
> 3. save the event, then try clicking the Brave Talk URL in it. instead of going to the URL when clicked, Google Calendar then opens a Google Maps link with the URL as the map location, which leaks the URL to Google Maps

I don't think that's a bug, per se. That behavior is exactly what happens with the Jitsi Meet extension for Google Calendar. Every meeting that I've been invited to has the meeting link in the location URL field. It is only google calendar that takes you to google maps if you click on the link. every other calendar app that i've used sees that it's a URL and opens it directly.

ps: appending the space does seem like a bug though (-;

diracdeltas commented:
@mrose17 it does seem like a bug because without the space it opens the link properly instead of in maps

mrose17 commented:
_oh!_ i was planning to get rid of space anyway, but now doubly so... thanks!

mrose17 commented:
> _oh!_ i was planning to get rid of space anyway, but now doubly so... thanks!

I see where the space is added, and it appears to be intentional. i have a theory as to why, but plan on checking with @tackley to confirm before making the change to https://github.com/brave/jidesha/blob/gt/tidy-and-use-webpack/src/meet-calendar.ts#L454

mrose17 commented on src/meet-calendar.ts:
```diff
@@ -0,0 +1,684 @@
+      // clicking save
+      window.setTimeout(function () {
+        locationNode.focus();
+        elem.val(elem.val() + " ");
```
I tried another 5 variants of this code, my favorite being replacing `+ " "` with `|| " "`, but i can't make it any more reliable. i am wondering whether we should use the `dispatchEvent` code  at https://github.com/brave/jidesha/blob/master/calendar/meet-calendar.js#L517-L525

tackley replied:
addressed by later updates

tackley commented:
In order to achieve testability and clarity on what the scripts in this extension are actually doing, relevant to our use case, I'm afraid I've had to give this yet more of an overhaul.

1.  I've extracted the gcal interactions in to `google-calendar.ts`, which comes with its own set of tests. These are really just smoke tests: they validate key interactions using jdom based on stub template html originally extracted from gcal. This is better than nothing, and at least provides a framework for future work.

2. With a combination of those tests and interaction on gcal itself, I went through every line of code and dom manipulation that was occurring to understand its exact purpose. This means that `google-calendar.ts` includes only what we actually need, and I've tried to structure it and add comments where needed to make it clear what is happening and why.

Apologies but this likely means that a full re-review of this PR will be necessary.

tackley commented:
Note also: I've updated the logic to set location, no longer requiring a "space" to be added addressing the issue @diracdeltas noted above. This works "for me" - retesting would be appreciated.

https://github.com/brave/jidesha/pull/2/files#diff-41d1d75547aaa966670df5187373362125bf37ce3334dc4010c4c5eb0d498101R90-R107

diracdeltas commented on src/google-calendar.ts:
```diff
@@ -0,0 +1,250 @@
+
+    await new Promise((resolve) => setTimeout(resolve, 200));
+
+    document.execCommand("insertHTML", false, newValue);
```
could you elaborate on why we are using execCommand? https://developer.mozilla.org/en-US/docs/Web/API/Document/execCommand says it is deprecated

also this calls insertHTML but looks like the newValue is just text - can it insert as text to avoid potential XSS vectors in newValue?

tackley replied:
Good suggestion, updated to `insertText` instead.

In terms of using execCommand, yes this is deprecated, but hours of trying alternative approaches did not find one that worked reliably: we need not just to update the content of the input box, but do it in such a way to causes the gcal javascript to process the entered input. Otherwise, the entry does not save when the meeting is saved.

As the MDN article notes:

> The Clipboard API can be used instead of execCommand in many cases, but execCommand is still sometimes useful. In particular, the Clipboard API doesn't replace the insertText command, which you can use to programmatically replace text at the cursor while preserving the undo buffer (edit history) in plain textarea and input elements.
